### PR TITLE
Get all serialization and REST tests passing

### PR DESCRIPTION
**Pull request by tdenniston**
Besides refactoring/code style updates, the main substantive change is the addition of the `SerializationState` field in the `Query` class. This allows us to allocate attribute data buffers and set them on the query objects without leaking memory/passing around invalid pointers.

All the serialization and REST tests now pass locally for me again.

@Shelnutt2 you are welcome to review this but it's a bit large so it's ok if you don't.

### CODE REVIEW

File: tiledb/rest/capnp/query.cc [hunk 1]
```diff
@@ -84,62 +77,64 @@ tiledb::sm::Status query_serialize(
         kj::Array<::capnp::word> protomessage = messageToFlatArray(message);
         kj::ArrayPtr<const char> message_chars = protomessage.asChars();
 
-        const auto nbytes = message_chars.size() + totalNumOfBytesInBuffers +
-                            totalNumOfBytesInOffsets +
-                            totalNumOfBytesInVarBuffers;
-        RETURN_NOT_OK(serialized_buffer->realloc(nbytes));
+        // Write the buffer headers
```
Comment (Shelnutt2) [inline tiledb/rest/capnp/query.cc]: Isn't this writing the entire query object not just the buffer headers?

File: tiledb/sm/query/query.cc [hunk 2]
```diff
@@ -281,404 +281,312 @@ Status Query::capnp(rest::capnp::Query::Builder* queryBuilder) {
...
-        RETURN_NOT_OK(
-            set_buffer(attributes[i], buff.buffer_, buff.buffer_size_))
-      }
+  // Serialize attribute buffer metadata
+  const auto attr_names = attributes();
+  auto attr_buffers_builder =
+      queryBuilder->initAttributeBufferHeaders(attr_names.size());
+  uint64_t total_fixed_len_bytes = 0;
+  uint64_t total_var_len_bytes = 0;
+  for (uint64_t i = 0; i < attr_names.size(); i++) {
```
Comment (Shelnutt2) [inline tiledb/sm/query/query.cc:339]: Just to validate, attributes() returns the vector of names based on the user set buffers (query.set_buffer calls), so `__coords` will also be included if set right?

File: tiledb/rest/capnp/query.cc [hunk 3]
```diff
@@ -52,22 +53,14 @@ tiledb::sm::Status query_serialize(
...
     serialized_buffer->reset_offset();
 
-    uint64_t totalNumOfBytesInBuffers{
-        query_builder.getTotalNumOfBytesInBuffers()};
-    uint64_t totalNumOfBytesInOffsets{
-        query_builder.getTotalNumOfBytesInOffsets()};
-    uint64_t totalNumOfBytesInVarBuffers{
-        query_builder.getTotalNumOfBytesInVarBuffers()};
-
+    uint64_t total_fixed_len_bytes =
```
Comment (Shelnutt2) [inline tiledb/rest/capnp/query.cc:61]: Thinking on this do we need to track the length of buffers separately or could this just be total_buffer_length_bytes?

**Review by Shelnutt2 — COMMENTED** [review]
So the `SerializationState` is going to be used for read queries, so internal tiledb can track the buffers and then when the query object is freed the buffers can also be freed?

File: tiledb/rest/capnp/query.cc (hunk 1 above)
Comment (tdenniston) [reply tiledb/rest/capnp/query.cc]: Yes, you are right -- I'll update the comment.

File: tiledb/rest/capnp/query.cc (hunk 3 above)
Comment (tdenniston) [reply tiledb/rest/capnp/query.cc:61]: I don't think we need these two values to be a part of the serialization process at all -- they are redundant (can be removed in a later PR).

File: tiledb/sm/query/query.cc (hunk 2 above)
Comment (tdenniston) [reply tiledb/sm/query/query.cc:339]: Yes.

**Comment by tdenniston** [comment]
> So the `SerializationState` is going to be used for read queries, so internal tiledb can track the buffers and then when the query object is freed the buffers can also be freed?

That's right -- because the `Buffer` objects are owned by the `Query` itself, they will be freed when the query object is destroyed (e.g. via `tiledb_query_free()`). That means we're free to allocate buffers internally (eventually it will only be for write queries, currently it's both read and write see #34 ) and use them for the query object.

**Pull request merged**